### PR TITLE
Handle `<style>` and `<pre>` tags in Handlebars/Glimmer

### PR DESCRIPTION
**Pull request by jurgenwerk**
## Description

This PR addresses two bugs in Handlebars/Glimmer printer:

1. Content in `<style>` tags is improperly formatted (everything turned into a one liner)
2. Same for `<pre>` tags (whose content should be left as is)

The solution to these bugs is to use the CSS parser for style tags by using [embed](https://prettier.io/docs/en/plugins.html#optional-embed) functionality, and completely ignore text processing for `<pre>` tags.

This also fixes issue https://github.com/prettier/prettier/issues/14261. 

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by jurgenwerk** [comment]
@fisker possible to review this one?

**Comment by jurgenwerk** [comment]
@fisker possible to review this one?

**Comment by oshmoun** [comment]
Many thanks for fixing this issue. Just started using prettier to format handlebars and wondered why this weird behaviour is occurring. Would be very great if this gets merged soon :pray:

**Comment by kachkaev** [comment]
I wonder if it’s also possible to format text inside `<script>` in addition to `<style>`. Then this PR can close three issues:
- https://github.com/prettier/prettier/issues/11463
- https://github.com/prettier/prettier/issues/14261
- https://github.com/prettier/prettier/issues/15685

**Review by NullVoxPopuli — APPROVED** [review]

File: src/language-handlebars/printer-glimmer.js [hunk 1]
```diff
@@ -764,11 +770,67 @@ function printBlockParams(node) {
...
+  return options.originalText.slice(
+    node.children.at(0).loc.start.offset,
+    node.children.at(-1).loc.end.offset,
+  );
+}
+
+function embed(path, options) {
+  const { node } = path;
+
+  if (node.type === "ElementNode") {
```
Comment (fisker) [inline src/language-handlebars/printer-glimmer.js:794]: I think we should check `TextNode` instead, `<style>` should not printed here.

**Comment by fisker** [comment]
I'm sorry that I missed this.

**Comment by fisker** [comment]
Since we have `--embedded-language-formatting=off`, we need add fallback in the generic print function. 

**Prettier pr-15087**
[Playground link](https://deploy-preview-15087--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAeAzjAngGzgPgB0oACEgCzgEMATOAJxOBOLLJoEt0AHHKrJCQBmeAB4BuFqTYArAK6YOQrAFpIsBDEE8qYOCoBGcGAHc4CSazYHdAawDm9CHKg1B9ewYAUAFgAMADT+QX4AlJZQVgC+UmwUAIxMsXEkALZUHhxQgn6SkDgQ9IIm5BzwEXExxKgA9Ji4BCABIBDcMBzQ6MigGU4mAAoZCF0oVDgm-F3NBvR2xgDKVKlwADJZcMhCY+hw07Ngtgvculn2yDD0crsgO6kc55fX6Kd4AIpyEPCb29cy6KLzF5wd6fDZILY4HbNACOH3g-Sc3BGICo6BUUHMdBoTRAFyoHBwpwAwhBUulkCicDgcc8oPY8ABBGAXDgGOTwhhrDHfSHXcgwVI4ADqpXgOj082GZQ4ADcylgKWB0FMQDKrgBJVyaeZgegcNoM1zzbB4HlQkDcJw7IWzbgUy1wHb0GUbZpZJ0wBFUezk8E-ZrHehOin0jhkhg4y1ZGBCjg0GDkZAADkCIHocFhHHTXp9VDN1xgVAMsfjiaQACZmgo4AAVIsjCHmuCpIw0LErKh0uTeuAAMUK6WZpwpECEQhAUSiQA)

```sh
--parser glimmer
--embedded-language-formatting off
```

**Input:**

```hbs
<style>
  header { 
    display: flex; 
    justify-content: space-between; 
    background: rgb(40,40,40); 

  } 
    h1 { 
      margin: 0; color: white; 
    } 
</style>
```

**Output:**

```hbs
<style>
  header { display: flex; justify-content: space-between; background:
  rgb(40,40,40); } h1 { margin: 0; color: white; }
</style>
```

**Comment by wycats** [comment]
What's blocking this?

File: src/language-handlebars/printer-glimmer.js (hunk 1 above)
Comment (jurgenwerk) [reply src/language-handlebars/printer-glimmer.js:794]: I'm not sure why, can you explain? The CSS content from `style` tags comes as `TextNode` in the `embed` function.

**Comment by jurgenwerk** [comment]
@fisker can you advise how that fallback would look like? I realize `embed` will be skipped in that case, but then `print` will handle CSS from your example's `<style>` content by formatting it - not quite sure how to skip formatting in this case.

**Comment by NullVoxPopuli** [comment]
Really eager to see this landed. So many style tags became unreadable with today's prettier

**Comment by jurgenwerk** [comment]
Apologies - I would need some clarifications from @fisker to address his comments. We have been using this branch at work where we are using style tags in single file components and it works without problems.

**Comment by cah-brian-gantzler** [comment]
Any update on this?

**Comment by NullVoxPopuli** [comment]
is anything blocking this work? it's unclear from comments / labels / etc

**Comment by jurgenwerk** [comment]
I posted a couple of responses to @fisker's review but there was no activity after that. This PR is now pretty old and has merge conflicts. I'll try to find some time to update it with upstream and ask for a review again.

File: src/language-handlebars/printer-glimmer.js [hunk 2]
```diff
@@ -156,6 +158,18 @@ function print(path, options, print) {
...
+      // 1.in <pre>,
+      // 2.in <style> tags when there are CSS syntax errors
+      // 3.when `embeddedLanguageFormatting` is disabled
+      if (
+        path.parent.tag === "pre" ||
+        path.parent.tag === "style" ||
+        options.embeddedLanguageFormatting === "off"
+      ) {
+        return node.chars;
+      }
```
Comment (fisker) [inline src/language-handlebars/printer-glimmer.js:171]: This is not necessary, if `embeddedLanguageFormatting: 'off'`, we should use the old logic.

File: src/language-handlebars/printer-glimmer.js (hunk 1 above)
Comment (fisker) [reply src/language-handlebars/printer-glimmer.js:810]: Because we may have tag print logic in the "generic" print. We should only care about CSS code.

**Comment by fisker** [comment]
> is anything blocking this work? it's unclear from comments / labels / etc

Sorry for the long delay, I was busy.

File: src/language-handlebars/printer-glimmer.js (hunk 1 above)
Comment (jurgenwerk) [reply src/language-handlebars/printer-glimmer.js]: Fixed in https://github.com/prettier/prettier/pull/15087/commits/6804b38b0c4d271758b95d97284e9615de37fe2d

**Comment by jurgenwerk** [comment]
@fisker I made the suggested changes, can you take another look?

File: src/language-handlebars/printer-glimmer.js (hunk 2 above)
Comment (jurgenwerk) [reply src/language-handlebars/printer-glimmer.js]: Fixed in https://github.com/prettier/prettier/pull/15087/commits/6804b38b0c4d271758b95d97284e9615de37fe2d

File: tests/format/handlebars/style-tag/__snapshots__/format.test.js.snap [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+<div>
+  <style>
+    .a {
+      color: red;
+    }
+  </style>
+</div>
+
+<div>
+  <style
```
Comment (jurgenwerk) [inline tests/format/handlebars/style-tag/__snapshots__/format.test.js.snap:206]: Need to fix this, `<style>` should not jump to 
`<style
>`

**Comment by jurgenwerk** [comment]
@fisker I made the suggested changes, can you take another look?

**Comment by fisker** [comment]
Test failure should related to https://github.com/nodejs/node/issues/54428

**Comment by wycats** [comment]
I'm really hoping we can make this happen because it's the primary issue blocking broader adoption of [glimmer-scoped-css](https://www.npmjs.com/package/glimmer-scoped-css/)

**Comment by jurgenwerk** [comment]
@fisker since test runner has issues with node 22.6.0 maybe we should try to upgrade to 22.7.0? Or rollback https://github.com/prettier/prettier/pull/16601?

**Comment by NullVoxPopuli** [comment]
Super excited for this!

**Comment by jurgenwerk** [comment]
@fisker since https://github.com/nodejs/node/issues/54428 is currently not getting attention, is there any chance we could address the node version related to the test runner failure? Ember community would really like to get this merged in as it's of paramount importance for https://github.com/cardstack/glimmer-scoped-css

**Comment by fisker** [comment]
The test failure is not related to Node.js, I'll fix.

**Comment by wycats** [comment]
So so excited for this to merge. Thanks so much @fisker!

**Comment by jurgenwerk** [comment]
@fisker can we merge this?

**Comment by NullVoxPopuli** [comment]
CI is green!

Very eager to start using this in my apps, so my `<style>` tags don't compress to a single line. (It's a real bear to edit things because today's behavior treats `<style>` as text and makes it all one line)

**Review by NullVoxPopuli — APPROVED** [review]

**Review by fisker — APPROVED** [review]

**Comment by jurgenwerk** [comment]
@fisker can we get this merged now that it's approved?

**Comment by NullVoxPopuli** [comment]
I'd love to be able to close this tab :D haha

**Comment by evoactivity** [comment]
I went digging into code and google searches looking for what I'd need to do to fix this then found this issue, would really love to see this land asap. My components cry out for multi line css!

**Comment by jurgenwerk** [comment]
@sosukesuzuki could you help to merge this? Thanks!

**Comment by NullVoxPopuli** [comment]
I suppose in the meantime, we can do:
```
"pnpm": {
  "overrides": {
    "prettier": "github:jurgenwerk/prettier#glimmer-embed-style-wip"
  }
}
```

in our package.jsons

**Comment by NullVoxPopuli** [comment]
using this on new projects, and it seems solid.

Can we merge?/release? <3

thank you!!

**Comment by jurgenwerk** [comment]
We've been using it successfully too. Does anyone know how to get in touch with the maintainers @fisker / @sosukesuzuki so they can merge this? Everything is ready and tested - we just need someone to press the button.

**Comment by fisker** [comment]
@jurgenwerk Use this https://github.com/prettier/prettier/pull/15087#issuecomment-2522427091 if you can't wait.

**Comment by jurgenwerk** [comment]
We have been waiting for a long time (the original PR dates back to 2023) so it's not a problem if we wait some more. I just wanted to make sure this PR does not get forgotten. Thanks for including this now. 👍

**Review by driesdl — APPROVED** [review]

**Review by NullVoxPopuli — APPROVED** [review]

**Comment by NullVoxPopuli** [comment]
@fisker thanks for setting up pkg-pr-new.

Really looking forward to this being in the next release of prettier. I appreciate you!!

**Pull request merged**